### PR TITLE
Make StrategicPatch delete all matching maps in a merging list

### LINKED ISSUE #38332: StrategicPatch should delete all duplicates from a list of maps
[StrategicPatch](https://github.com/kubernetes/kubernetes/blob/v1.6.0-alpha.0/pkg/util/strategicpatch/patch.go#L659) should delete all duplicates from a list of maps.
Currently it just delete the first one appears in the list.

We should make StrategicPatch delete all duplicates no matter it is a list of maps or a list of primitives. (StrategicPatch don't support list of lists).

Ref : https://github.com/kubernetes/community/pull/140#issuecomment-265598220

cc: @lavalamp

### PR DESCRIPTION
**Pull request by mengqiy**
fixes #38332

```release-note
NONE
```

cc: @lavalamp @pwittrock

### CODE REVIEW

**Comment by ymqytw** [comment]
@kubernetes/sig-api-machinery-misc Anyone help to review this small PR?

File: pkg/util/strategicpatch/patch.go [hunk 1]
```diff
@@ -656,12 +656,16 @@ func mergeSlice(original, patch []interface{}, elemType reflect.Type, mergeKey s
 			if patchType == deleteDirective {
 				mergeValue, ok := typedV[mergeKey]
 				if ok {
-					_, originalKey, found, err := findMapInSliceBasedOnKeyValue(original, mergeKey, mergeValue)
-					if err != nil {
-						return nil, err
-					}
-
-					if found {
+					// delete all matching maps (based on merge key) from a merging list
```
Comment (lavalamp) [inline pkg/util/strategicpatch/patch.go]: s/maps/entries/

File: pkg/util/strategicpatch/patch_test.go [hunk 2]
```diff
@@ -308,6 +308,25 @@ testCases:
       $patch: replace
     modified:
       other: a
+  - description: delete all duplicate maps in a merging list
```
Comment (lavalamp) [inline pkg/util/strategicpatch/patch_test.go]: s/maps/entries/

**Comment by lavalamp** [comment]
one word change, then LGTM

**Comment by ymqytw** [comment]
@lavalamp Done. Thanks for reviewing.

**Comment by ymqytw** [comment]
@k8s-bot bazel test this

**Comment by ymqytw** [comment]
@k8s-bot unit test this

**Review by lavalamp — APPROVED** [review]
/lgtm

**Pull request merged**